token/transfer-hook-interface/src/instruction.rs

```diff
@@ -0,0 +1,206 @@
+    const TLV_DISCRIMINATOR: Discriminator = Discriminator::new(EXECUTE_DISCRIMINATOR);
+}
+/// First 8 bytes of `hash::hashv(&["spl-transfer-hook-interface:execute"])`
+const EXECUTE_DISCRIMINATOR: [u8; Discriminator::LENGTH] = [105, 37, 101, 197, 75, 251, 102, 26];
+// annoying, but needed to perform a match on the value
+const EXECUTE_DISCRIMINATOR_SLICE: &[u8] = &EXECUTE_DISCRIMINATOR;
+/// First 8 bytes of `hash::hashv(&["spl-transfer-hook-interface:initialize-extra-account-metas"])`
+const INITIALIZE_EXTRA_ACCOUNT_METAS_DISCRIMINATOR: &[u8] = &[43, 34, 13, 49, 167, 88, 235, 235];
+
+impl TransferHookInstruction {
+    /// Unpacks a byte buffer into a [TransferHookInstruction](enum.TransferHookInstruction.html).
+    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
+        if input.len() < Discriminator::LENGTH {
+            return Err(ProgramError::InvalidInstructionData);
+        }
+        let (discriminator, rest) = input.split_at(Discriminator::LENGTH);
+        Ok(match discriminator {
+            EXECUTE_DISCRIMINATOR_SLICE => {
+                let amount = rest
+                    .get(..8)
+                    .and_then(|slice| slice.try_into().ok())
+                    .map(u64::from_le_bytes)
+                    .ok_or(ProgramError::InvalidInstructionData)?;
+                Self::Execute { amount }
+            }
+            INITIALIZE_EXTRA_ACCOUNT_METAS_DISCRIMINATOR => Self::InitializeExtraAccountMetas,
+            _ => return Err(ProgramError::InvalidInstructionData),
+        })
+    }
+
+    /// Packs a [TokenInstruction](enum.TokenInstruction.html) into a byte buffer.
+    pub fn pack(&self) -> Vec<u8> {
+        let mut buf = vec![];
+        match self {
+            Self::Execute { amount } => {
+                buf.extend_from_slice(EXECUTE_DISCRIMINATOR_SLICE);
+                buf.extend_from_slice(&amount.to_le_bytes());
+            }
+            Self::InitializeExtraAccountMetas => {
+                buf.extend_from_slice(INITIALIZE_EXTRA_ACCOUNT_METAS_DISCRIMINATOR);
+            }
+        };
+        buf
+    }
+}
+
+/// Creates an `Execute` instruction, provided all of the additional required
+/// account metas
+#[allow(clippy::too_many_arguments)]
+pub fn execute_with_extra_account_metas(
+    program_id: &Pubkey,
+    source_pubkey: &Pubkey,
+    mint_pubkey: &Pubkey,
+    destination_pubkey: &Pubkey,
+    authority_pubkey: &Pubkey,
+    validate_state_pubkey: &Pubkey,
+    additional_accounts: &[AccountMeta],
+    amount: u64,
+) -> Instruction {
+    let mut instruction = execute(
+        program_id,
+        source_pubkey,
+        mint_pubkey,
+        destination_pubkey,
+        authority_pubkey,
+        validate_state_pubkey,
+        amount,
+    );
+    instruction.accounts.extend_from_slice(additional_accounts);
+    instruction
+}
+
+/// Creates an `Execute` instruction, without the additional accounts
+#[allow(clippy::too_many_arguments)]
+pub fn execute(
```

Comment:
Can this be renamed to `create_execute_instruction`? `execute` sounds like its making the `invoke` call

Reply:
it's not particularly consistent since none of the Rust crates follow the `create_<ix_name>_instruction`. Maybe we should rename `execute` to `perform_transfer_hook` or something?

Reply:
Maybe instead rename the CPI helper from `execute<'a>` to `invoke_hook<'a>` or `hook_invoke<'a>`?

Reply:
Agreed with Noah though it's a bit confusing since the functions return `Instruction`, so might I suggest `build_execute_instruction` or something like that?

Reply:
Or `create_` hah (Noah's comment)

Reply:
Everywhere else in programs, it's `crate::instruction::instruction_name`, which I've always liked, since it's very clear and avoids repetition.

It sounds like `Execute` is probably a bad name. Maybe `HandleTransfer` would be better? Or `ProcessTransfer`? Or `CallProcedure`? Or `HandleEvent`?

Reply:
I think `Execute` is okay, especially since it's actual ix-bytes are hashed with the prefix of the interface name.

👍 lgtm